### ISSUE TITLE
AlertmanagerSpec configSecret only mounts `alertmanager.yaml`

### ISSUE THREAD

**Issue by joshbranham** [prometheus-operator/prometheus-operator#3638]
**What happened?**

Utilized `configSecret` field of the `AlertmanagerSpec` to mount 2 files to `/etc/alertmanager/config`, however only the key named `alertmanager.yaml` was mounted.

**Did you expect to see something different?**

I expected both files to be mounted (in this case it was `.tmpl` file for notification templates)

**How to reproduce it (as minimally and precisely as possible)**:

Create a secret with two file keys:

```
apiVersion: v1
data:
  alertmanager.yaml: <some base64>
  my.tmpl: <some base64>
kind: Secret
metadata:
  annotations:
```

**Environment**

* Prometheus Operator version:

    `prometheus-operator:v0.43.0`
    

* Kubernetes version information:

    `v1.18.1`
    

* Kubernetes cluster kind:

    kops

* Manifests:

```
apiVersion: monitoring.coreos.com/v1
kind: Alertmanager
metadata:
  labels:
    alertmanager: main
  name: main
  namespace: monitoring
spec:
  configSecret: alertmanager-config-8hcb55hchg < secret with two files
  externalUrl: *********
  image: quay.io/prometheus/alertmanager
  nodeSelector:
    kubernetes.io/os: linux
  replicas: 3
  securityContext:
    fsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
  serviceAccountName: alertmanager-main
  version: v0.21.0
```

* Prometheus Operator Logs:

```
insert Prometheus Operator logs relevant to the issue here
```

**Anything else we need to know?**:

My guess is the code is explicitly looking for the `alertmanager.yaml` key and nothing else, which is fine. What I was trying to do was use `kustomize` to `configMapGenerator/secretGenerator` to take my alertmanager templates and not only mount them to a known path, but let `kustomize` use a sha on the name so alertmanager also reloads when a new one is set. Lastly, this does not work because in your alertmanager config, you can't wildcard a directory in the `templates` path (example below)

```
global:
  resolve_timeout: 5m
  slack_api_url: https://SLACK_TOKEN
templates:
- '/etc/alertmanager/configmaps/alertmanager-templates-SOMESHA?/my.tmpl'
```

**Comment by desaintmartin** [comment]
Note that it happens in 0.43.0 even without configSecret, only by having the "default" secret with several keys.

**Comment by joshbranham** [comment]
> Note that it happens in 0.43.0 even without configSecret, by having the "default" `alertmanager-{ALERTMANAGER_NAME}` secret with several keys, according to https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/alerting.md#manually-managed-secret

Interesting, in this case will the sidecar reload on template changes and `alertmanager.yaml` changes?

**Comment by simonpasquier** [comment]
I see why this is happening now. Since the introduction of `AlertmanagerConfig` CRD, the user-provided secret containing the Alertmanager configuration isn't mounted directly in the container. Instead the operator will merge this "raw" Alertmanager configuration with all `AlertmanagerConfig` objects and generate a new managed secret containing only the `alertmanager.yaml` key.

A workaround for now is to define the templates in a separate secret, add it to the additional Alertmanager's secrets and modify the `templates` path to be `/etc/alertmanager/secrets/`.

The fix would be to copy all keys from the user-provided secret into the managed secret.

**Comment by desaintmartin** [comment]
> > Note that it happens in 0.43.0 even without configSecret, by having the "default" `alertmanager-{ALERTMANAGER_NAME}` secret with several keys, according to https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/alerting.md#manually-managed-secret
> 
> Interesting, in this case will the sidecar reload on template changes and `alertmanager.yaml` changes?

Yes, it is. alertmanager.yaml part is working as intended and changes are propagated to `alertmanager-{ALERTMANAGER_NAME}-generated` secret, and AlertManager correctly reloads.

**Comment by desaintmartin** [comment]
> The fix would be to copy all keys from the user-provided secret into the managed secret.

Unfortunately, prometheus-operator actively converge, if one edits the managed secret, to the "expected" result.

**Comment by simonpasquier** [comment]
> Unfortunately, prometheus-operator actively converges, if one edits the managed secret, to the "expected" result.

I meant a fix in the operator code which I'm working on :)

**Comment by simonpasquier** [comment]
Closed by #3647
